src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexFindOptimizations.cs

```diff
@@ -27,28 +28,48 @@ public RegexFindOptimizations(RegexTree tree, CultureInfo culture)
-                LeadingAnchor &= ~RegexPrefixAnalyzer.Bol;
+                LeadingAnchor = RegexNodeKind.Unknown;
             }
-            if ((LeadingAnchor & (RegexPrefixAnalyzer.Beginning | RegexPrefixAnalyzer.Start | RegexPrefixAnalyzer.EndZ | RegexPrefixAnalyzer.End)) != 0)
+            if (LeadingAnchor is RegexNodeKind.Beginning or RegexNodeKind.Start or RegexNodeKind.EndZ or RegexNodeKind.End)
             {
                 FindMode = (LeadingAnchor, _rightToLeft) switch
                 {
-                    (RegexPrefixAnalyzer.Beginning, false) => FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Beginning,
-                    (RegexPrefixAnalyzer.Beginning, true) => FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Beginning,
-                    (RegexPrefixAnalyzer.Start, false) => FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Start,
-                    (RegexPrefixAnalyzer.Start, true) => FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Start,
-                    (RegexPrefixAnalyzer.End, false) => FindNextStartingPositionMode.LeadingAnchor_LeftToRight_End,
-                    (RegexPrefixAnalyzer.End, true) => FindNextStartingPositionMode.LeadingAnchor_RightToLeft_End,
+                    (RegexNodeKind.Beginning, false) => FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Beginning,
+                    (RegexNodeKind.Beginning, true) => FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Beginning,
+                    (RegexNodeKind.Start, false) => FindNextStartingPositionMode.LeadingAnchor_LeftToRight_Start,
+                    (RegexNodeKind.Start, true) => FindNextStartingPositionMode.LeadingAnchor_RightToLeft_Start,
+                    (RegexNodeKind.End, false) => FindNextStartingPositionMode.LeadingAnchor_LeftToRight_End,
+                    (RegexNodeKind.End, true) => FindNextStartingPositionMode.LeadingAnchor_RightToLeft_End,
                     (_, false) => FindNextStartingPositionMode.LeadingAnchor_LeftToRight_EndZ,
                     (_, true) => FindNextStartingPositionMode.LeadingAnchor_RightToLeft_EndZ,
                 };
                 return;
             }
 
+            // Compute any anchor trailing the expression.  If there is one, and we can also compute a fixed length
+            // for the whole expression, we can use that to quickly jump to the right location in the input.
+            if (!_rightToLeft) // haven't added FindNextStartingPositionMode support for RTL
+            {
+                TrailingAnchor = RegexPrefixAnalyzer.FindTrailingAnchor(tree.Root);
+                if (TrailingAnchor is RegexNodeKind.End or RegexNodeKind.EndZ &&
+                    tree.Root.ComputeMaxLength() is int maxLength)
+                {
+                    Debug.Assert(maxLength >= _minRequiredLength, $"{maxLength} should have been greater than {_minRequiredLength} minimum");
+                    MaxPossibleLength = maxLength;
```

Comment:
Oh I see, you only save the maximum possible length in the case of trailing anchors and the optimization is to jump to that directly. I wonder if we have considered instead of using a single MaxPossibleLength for this optimization, to use a List/array/collection withall possible starting positions given the diffferent lengths of the alternation, and that might help making FindFirstChar more efficient for some cases where the longest branch isn't the one that should match the input.

Of course, this suggestion doesn't have to be added to the PR, just thinking out loud to see if this is something worth adding later.

Reply:
> I wonder if we have considered instead of using a single MaxPossibleLength for this optimization, to use a List/array/collection withall possible starting positions given the diffferent lengths of the alternation, and that might help making FindFirstChar more efficient for some cases where the longest branch isn't the one that should match the input.

I don't know that it would help FindFirstChar.  But I did make the computation based off of any RegexNode rather than off of RegexTree with the idea that we could then use it for individual nodes during code generation where it makes sense.  For example, in support of https://github.com/dotnet/runtime/issues/64097, if we're generating code for a top-level alternation like `^something|else$`, we can emit a specialized implementation for it that first processes the first branch, and then if that doesn't match, jumps to four before the end to process the second branch, which we'd compute using ComputeMaxLength on the individual branch.